client/modules/sheaf-core/src/index.ts

```diff
@@ -0,0 +1,204 @@
+  drawSelection,
+  keymap
+} from "@codemirror/view"
+
+import { highlightActiveLine, highlightSpecialChars } from "@codemirror/view"
+import { history, historyKeymap } from "@codemirror/history"
+import { foldGutter, foldKeymap } from "@codemirror/fold"
+import { syntaxTree, indentOnInput } from "@codemirror/language"
+import { lineNumbers } from "@codemirror/gutter"
+import { defaultKeymap, defaultTabBinding } from "@codemirror/commands"
+import { bracketMatching } from "@codemirror/matchbrackets"
+import { closeBrackets, closeBracketsKeymap } from "@codemirror/closebrackets"
+import { highlightSelectionMatches, searchKeymap } from "@codemirror/search"
+import { autocompletion, completionKeymap } from "@codemirror/autocomplete"
+import { commentKeymap } from "@codemirror/comment"
+import { rectangularSelection } from "@codemirror/rectangular-selection"
+import { redo } from "@codemirror/history"
+import { copyLineDown } from "@codemirror/commands"
+import { nextDiagnostic, openLintPanel } from "@codemirror/lint"
+
+import { writable } from "svelte/store"
+
+import { debounce } from "wj-util"
+import { printTree } from "./print-tree"
+import { confinement } from "./theme"
+import { indentHack } from "./extensions/indent-hack"
+
+interface EditorStore {
+  /** The current document of the editor. */
+  doc: EditorState["doc"]
+  /** The current 'value' (content) of the editor. */
+  value: string
+}
+
+export class SheafCore {
+  /** The element the editor is attached to. */
+  parent!: Element
+
+  /** The CodeMirror `EditorState` the editor has currently.
+   *  The state is immutable and is replaced as the editor updates. */
+  state = EditorState.create()
+
+  /** The CodeMirror `EditorView` instance the editor interacts with the DOM with. */
+  view!: EditorView
+
+  /** A store that allows reactive access to editor state. */
+  store = writable<EditorStore>({ doc: this.state.doc, value: "" })
+  subscribe = this.store.subscribe
+  set = this.store.set
+
+  /** The lines currently being interacted with by the user.
+   *  This includes all selected lines, the line the cursor is present on, etc. */
+  activeLines = writable(new Set<number>())
+
+  private spellcheckCompartment = new Compartment()
+  private guttersCompartment = new Compartment()
+
+  /** Starts the editor. */
+  async init(parent: Element, doc: string, extensions: Extension[] = []) {
+    this.parent = parent
+
+    const updateState = debounce(() => this.refresh(), 50)
+
+    const updateHandler = ViewPlugin.define(() => ({
+      update: (update: ViewUpdate) => {
+        // update store on change
+        if (update.docChanged) updateState()
+        // get active lines
+        if (update.selectionSet || update.docChanged) {
+          const activeLines: Set<number> = new Set()
+          for (const r of update.state.selection.ranges) {
+            const lnStart = update.state.doc.lineAt(r.from).number
+            const lnEnd = update.state.doc.lineAt(r.to).number
```

Comment:
Just as a general comment, could we try to avoid unnecessarily-shortened variable names? It can be difficult to work out what they mean
```suggestion
          for (const range of update.state.selection.ranges) {
            const lnStart = update.state.doc.lineAt(range.from).number
            const lnEnd = update.state.doc.lineAt(range.to).number
```

Reply:
A bunch of this code is older, or comes from the internals of CodeMirror, but yes, I normally wouldn't do that. Also, I planned to clean this up a bunch anyways - I'd rather not be in the weeds renaming variables right now.